_posts/2024-11-22-faiss-byte-vector.md

```diff
@@ -0,0 +1,192 @@
+|**Dataset 3**	|**8**	|6.75	|5.98	|7.69	|7.7	|
+|**Dataset 3**	|**16**	|10.51	|6.94	|13.87	|12.4	|
+|	|	|	|	|	|	|
+|**Dataset 4**	|**1**	|2.91	|3.03	|3.16	|3.15	|
+|**Dataset 4**|**8**	|3.38	|3.30	|6.30	|4.75	|
+|**Dataset 4**	|**16**	|4.35	|3.80	|8.76	|8.83	|
+
+### Key findings
+
+When comparing the benchmarking results, here are the key findings:
+
+- **Memory savings**: Byte vectors reduced memory usage by up to **72%**, with higher-dimensional vectors achieving greater reductions.
+- **Indexing performance**: The mean indexing throughput for byte vectors was **2x to 107.84%** higher than for float vectors, especially with larger vector dimensions.
+- **Search performance**: Search latencies were similar, with byte vectors occasionally performing better.
+- **Recall**: For byte vectors, there was a slight (up to **8.8%**) reduction in recall compared to float vectors, depending on the dataset and the quantization technique used.
+
+## How does Faiss work with byte vectors internally?
+
+Faiss doesn't directly support the `byte` data type for storing vectors. To achieve this, OpenSearch uses a  [`QT_8bit_direct_signed` scalar quantizer](https://faiss.ai/cpp_api/struct/structfaiss_1_1ScalarQuantizer.html). This quantizer accepts float vectors within the signed 8-bit value range and encodes them as unsigned 8-bit integer vectors. During indexing and search, these encoded unsigned 8-bit integer vectors are decoded back into signed 8-bit original vectors for distance computation.
+
+This quantization approach reduces the memory footprint by a factor of four. However, encoding and decoding during scalar quantization introduce additional latency. To mitigate this, you can use [SIMD optimization](https://opensearch.org/docs/latest/search-plugins/knn/knn-index#simd-optimization-for-the-faiss-engine) with the `QT_8bit_direct_signed` quantizer to reduce search latencies and improve indexing throughput.
+
+### Example
+
+The following example shows how an input vector is encoded and decoded using `QT_8bit_direct_signed` scalar quantizer:
+
+```c
+// Input vector:
+[-126, 28, 127, 0, 10, -45, 12, -110]
+
+// Encoded vector generated by adding 128 to each dimension of the input vector to convert signed int8 to unsigned int8:
+[2, 156, 255, 128, 138, 83, 140, 18]
+
+// Encoded vector is decoded back into the original signed int8 vector by subtracting 128 from each dimension for distance computation:
+[-126, 28, 127, 0, 10, -45, 12, -110]
+```
+
+## Conclusion
+
+OpenSearch 2.17 introduced support for Faiss byte vectors, enabling you to store quantized byte vector embeddings efficiently. This reduces memory consumption by up to 75%, lowers costs, and maintains high performance. These advantages make byte vectors an excellent choice for large-scale similarity search applications, especially where memory resources are limited, and applications that handle large volumes of data within the signed byte value range.
+
+## Future enhancements
```

Comment:
Nit: This section should probably precede the conclusion.

Reply:
Done